plan/build_key_info.go

```diff
@@ -0,0 +1,215 @@
+			schema.Append(&expression.Column{
+				FromID:   "",
+				Position: -1})
+		}
+	}
+	for _, key := range p.GetChildren()[0].GetSchema().Keys {
+		ok := true
+		newKey := make([]*expression.Column, 0, len(key))
+		for _, keyCol := range key {
+			pos := schema.GetColumnIndex(keyCol)
+			if pos == -1 {
+				ok = false
+				break
+			} else {
+				newKey = append(newKey, p.schema.Columns[pos])
+			}
+		}
+		if ok {
+			p.schema.Keys = append(p.schema.Keys, newKey)
+		}
+	}
+}
+
+func (p *Projection) buildKeyInfo() {
+	p.baseLogicalPlan.buildKeyInfo()
+	// Projection's schema is create from expression of projection.
+	// And the ith column in schema correspond to the ith expression.
+	// So we create a temp schema from p.Exprs to build key information.
+	schema := expression.NewSchema(make([]*expression.Column, 0, p.schema.Len()))
+	for _, expr := range p.Exprs {
+		if col, isCol := expr.(*expression.Column); isCol {
+			schema.Append(col)
+		} else {
+			// If the expression is not a column, we add a column to occupy the position.
+			schema.Append(&expression.Column{
+				FromID:   "",
+				Position: -1})
+		}
+	}
+	for _, key := range p.GetChildren()[0].GetSchema().Keys {
+		ok := true
+		newKey := make([]*expression.Column, 0, len(key))
+		for _, keyCol := range key {
+			pos := schema.GetColumnIndex(keyCol)
+			if pos == -1 {
+				ok = false
+				break
+			} else {
+				newKey = append(newKey, p.schema.Columns[pos])
+			}
+		}
+		if ok {
+			p.schema.Keys = append(p.schema.Keys, newKey)
+		}
+	}
+}
+
+func (p *Trim) buildKeyInfo() {
+	p.baseLogicalPlan.buildKeyInfo()
+	for _, key := range p.children[0].GetSchema().Keys {
+		ok := false
+		newKey := make([]*expression.Column, 0, len(key))
+		for _, col := range key {
+			pos := p.schema.GetColumnIndex(col)
+			if pos == -1 {
+				ok = false
+				break
+			}
+			newKey = append(newKey, p.schema.Columns[pos])
+		}
+		if ok {
+			p.schema.Keys = append(p.schema.Keys, newKey)
+		}
+	}
+}
+
+func (p *Join) buildKeyInfo() {
+	p.baseLogicalPlan.buildKeyInfo()
+	switch p.JoinType {
+	case SemiJoin, SemiJoinWithAux:
+		p.schema.Keys = p.children[0].GetSchema().Clone().Keys
+	case InnerJoin, LeftOuterJoin, RightOuterJoin:
+		// We first check the equal conditions. If one of the equal conditions' both sides are unique key,
+		// all unique key information will be reserved. Otherwise none of them is unique key any more.
+		if len(p.EqualConditions) == 0 {
+			return
+		}
+		ok := false
+		for _, expr := range p.EqualConditions {
+			ln := expr.GetArgs()[0].(*expression.Column)
+			rn := expr.GetArgs()[1].(*expression.Column)
+			lOk := false
+			rOk := false
+			for _, key := range p.children[0].GetSchema().Keys {
+				if len(key) == 1 && key[0].Equal(ln, p.ctx) {
+					lOk = true
+					break
+				}
+			}
+			for _, key := range p.children[1].GetSchema().Keys {
+				if len(key) == 1 && key[0].Equal(rn, p.ctx) {
+					rOk = true
+					break
+				}
+			}
+			if lOk && rOk {
```

Comment:
if lOk, p.schema.keys contains the right child's key.
if rOk, p.schema.keys contains the left child's key.